src/main/java/org/apache/sling/models/impl/ModelAdapterFactory.java

```diff
@@ -495,12 +480,97 @@ public RuntimeException inject(InjectableElement element, Object value) {
         }
     }
 
+    @NotNull
+    private InjectableElement getElement(InjectableElement element, Type type) {
+        return new InjectableElement() {
+            @Override
+            public AnnotatedElement getAnnotatedElement() {
+                return element.getAnnotatedElement();
+            }
+
+            @Override
+            public Type getType() {
+                return type;
+            }
+
+            @Override
+            public boolean isPrimitive() {
+                return element.isPrimitive();
+            }
+
+            @Override
+            public String getName() {
+                return element.getName();
+            }
+
+            @Override
+            public String getSource() {
+                return element.getSource();
+            }
+
+            @Override
+            public String getVia() {
+                return element.getVia();
+            }
+
+            @Override
+            public Class<? extends ViaProviderType> getViaProviderType() {
+                return element.getViaProviderType();
+            }
+
+            @Override
+            public boolean hasDefaultValue() {
+                return element.hasDefaultValue();
+            }
+
+            @Override
+            public Object getDefaultValue() {
+                return element.getDefaultValue() == null ? Optional.empty() : element.getDefaultValue();
```

Comment:
Ok, after reading through this again (after making the comment below), I understand where the confusion comes from. Changing the type here is causing you to have some perplexing code in the `InjectCallback` below. Instead of handling Optional in two places, let the default value propagate `null` and use the null-safe wrapper method, `Optional.ofNullable(..)`.

Reply:
Good idea, this simplifies things a bit, will give it a try.

Reply:
Here, unfortunately, I have to provide a value which is not null, to make the injection successful as we don't have a specific injector for Optional type and a default value injection will be executed last and and will fail if the value is not present (null)